crates/storage/src/cache.rs

```diff
@@ -1,252 +1,150 @@
-/// keep the reservation to it.
-type CacheSlot<T> = Arc<RwLock<SlotState<T>>>;
-
-/// Describes a cache entry that may be occupied, reserved for pending database read, or returned an
-/// error from a database read.
-#[derive(Debug)]
-pub(crate) enum SlotState<T> {
-    /// Authentic database entry.
-    Ready(T),
-
-    /// A database fetch is happening in the background and it will be updated.
-    Pending(broadcast::Receiver<T>),
-
-    /// An unspecified error happened fetching from the database.
-    Error,
-}
-
-impl<T: Clone> SlotState<T> {
-    /// Tries to read a value from the slot, asynchronously.
-    pub(crate) async fn get_async(&self) -> DbResult<T> {
-        match self {
-            Self::Ready(v) => Ok(v.clone()),
-            Self::Pending(ch) => {
-                // When we see this log get triggered and but feels like the corresponding fetch is
-                // hanging for this read then it means that this code wasn't implemented
-                // correctly.
-                // TODO figure out how to test this
-                trace!("waiting for database fetch to complete");
-                match ch.resubscribe().recv().await {
-                    Ok(v) => Ok(v),
-                    Err(_e) => Err(DbError::WorkerFailedStrangely),
-                }
-            }
-            Self::Error => Err(DbError::CacheLoadFail),
-        }
-    }
-
-    /// Tries to read a value from the slot, blockingly.
-    pub(crate) fn get_blocking(&self) -> DbResult<T> {
-        match self {
-            Self::Ready(v) => Ok(v.clone()),
-            Self::Pending(ch) => {
-                // When we see this log get triggered and but feels like the corresponding fetch is
-                // hanging for this read then it means that this code wasn't implemented
-                // correctly.
-                // TODO figure out how to test this
-                trace!("waiting for database fetch to complete");
-                match ch.resubscribe().blocking_recv() {
-                    Ok(v) => Ok(v),
-                    Err(_e) => Err(DbError::WorkerFailedStrangely),
-                }
-            }
-            Self::Error => Err(DbError::CacheLoadFail),
-        }
-    }
-}
-
-/// Wrapper around a LRU cache that handles cache reservations and asynchronously waiting for
-/// database operations in the background without keeping a global lock on the cache.
 pub(crate) struct CacheTable<K, V> {
-    cache: Mutex<lru::LruCache<K, CacheSlot<V>>>,
+    /// The actual cache.
+    cache: RwLock<LruCache<K, V>>,
+
+    /// One mutex per key to prevent duplicate fetches.
+    /// Created when inserting, cleaned up when the entry is purged.
+    /// Not sure if trying to clean this up after fetch complete is worthwhile. Seems to add a
+    /// bunch of complex logic whereas keeping it won't have much memory footprint.
+    fetch_mutexes: RwLock<HashMap<K, Arc<Mutex<()>>>>,
```

Comment:
This is a bit of a red flag.  The broadcast channel was to allow subsequent queries for the same key to reuse the freshly fetched value from the cache.  This approach changes the flow quite a bit.

Reply:
It changes the flow, but the behavior is still the same.

Reply:
The behavior is not the same.  The locking pattern is different (more contention) and it introduces this new dummy lock structure.